### PR TITLE
Database Projects - Add existing file

### LINKED ISSUE #14383: Database Projects - Add existing file to project
**Is your feature request related to a problem? Please describe.**
Because it is a requirement, I need to add existing files into a database project (yeah I know it's not awesome but still it is there)

I can of course do this by editing the sqlproj file but this is unsuitable for some users.

**Describe the solution or feature you'd like**
Add Existing Item to be added to this menu and to the command palette

image

**Describe alternatives you've considered**
Manual editing of sqlproj file works but is unsuitable for all

Thank you very much.

### PR DESCRIPTION
**Pull request by zijchen**
This PR fixes #14383

image

In-line with Visual Studio behavior:
* SQL files will be automatically added to build `<Build Include="..." />`
* All other files will be added as `<None Include="..." />`

If the file has been previously excluded, the `<Build/None Remove="..." />` will be removed before adding the Include (if necessary). 

For SDK-style projects, if the SQL file falls under default glob, no `<Build Include="..." />` will be added.

### CODE REVIEW

**Comment by zijchen** [comment]
I'll add some steps to the test plans too after this is checked in

File: extensions/sql-database-projects/src/models/project.ts [hunk 1]
```diff
@@ -912,6 +942,10 @@ export class Project implements ISqlProject {
 	}
 
 	private async addFileToProjFile(filePath: string, xmlTag: string, attributes?: Map<string, string>): Promise<void> {
+
+		// determine if the file has been previously excluded
```
Comment (kisantia) [inline extensions/sql-database-projects/src/models/project.ts:946]: "delete Remove node if a file has been previously excluded"

**Review by kisantia — APPROVED** [review]

File: extensions/sql-database-projects/src/models/project.ts [hunk 2]
```diff
@@ -1029,6 +1063,33 @@ export class Project implements ISqlProject {
...
+	private async undoExcludeFileFromProjFile(xmlTag: string, relativePath: string): Promise<void> {
+		const nodes = this.projFileXmlDoc!.documentElement.getElementsByTagName(xmlTag);
+		for (let i = 0; i < nodes.length; i++) {
+			const parent = nodes[i].parentNode;
+			if (parent && nodes[i].getAttribute(constants.Remove) === utils.convertSlashesForSqlProj(relativePath)) {
+				parent.removeChild(nodes[i]);
+
+				// delete ItemGroup if this was the only entry
+				// only want element nodes, not text nodes
+				const otherChildren = Array.from(parent.childNodes).filter((c: any) => c.childNodes);
```
Comment (kisantia) [inline extensions/sql-database-projects/src/models/project.ts:1080]: can this reuse `removeNode()` to remove the Remove node?

Comment (Charles-Gagnon) [inline extensions/sql-database-projects/src/models/project.ts:1080]: childNodes should have type `ChildNode`, you should use that instead of `any`

Comment (kisantia) [reply extensions/sql-database-projects/src/models/project.ts:1080]: I sent a PR to update the rest of this file where any was being used #18091

**Review by kisantia — APPROVED** [review]

File: extensions/sql-database-projects/src/test/project.test.ts [hunk 3]
```diff
@@ -1358,6 +1390,51 @@ describe('Project: sdk style project content operations', function (): void {
 		should(project.projectGuid).not.equal(undefined);
 		should(projFileText.includes(constants.ProjectGuid)).equal(true);
 	});
+
+	it('Should handle adding existing items to project', async function (): Promise<void> {
+		projFilePath = await testUtils.createTestSqlProjFile(baselines.openSdkStyleSqlProjectBaseline);
+		const projectFolder = path.dirname(projFilePath);
+
+		// Create a sql file inside project root
```
[the remark this answers is not in the document]
Comment (zijchen) [reply extensions/sql-database-projects/src/test/project.test.ts:1438]: Added

File: extensions/sql-database-projects/src/models/project.ts (hunk 2 above)
Comment (zijchen) [reply extensions/sql-database-projects/src/models/project.ts]: Had to change the logic of removeNode() a bit but is re-using code now

**Comment by kisantia** [comment]
The "Should handle adding existing items to project" test is failing, so please make sure to fix it before checking in. Thanks!

**Comment by zijchen** [comment]
@kisantia the test failure is fixed. It came from the test case I added for adding existing file from outside the project root. There was nothing wrong with the source code, just had to tweak the test code a bit regarding relative paths and Linux specifics.

**Comment by kisantia** [comment]
> @kisantia the test failure is fixed. It came from the test case I added for adding existing file from outside the project root. There was nothing wrong with the source code, just had to tweak the test code a bit regarding relative paths and Linux specifics.

Awesome, glad it just needed a small fix. Should be good to check in then after the PR validation finishes!

**Pull request merged**